### DISCUSSION TITLE
Availability of Python types for SeleniumBase?

### DISCUSSION THREAD

**Discussion by albertkim** [seleniumbase/SeleniumBase#3239] [General]
I'm a big fan of my IDE knowing that types/functions are available as I'm coding. However, right from the start with code like

```
    with SB(uc=True, headless=True) as sb:
        sb.get(params["start_url"])
```

VSCode thinks that the `sb` type is `Unknown`, meaning that there is no auto-complete for any function names called on sb. I'm wondering if I'm doing something wrong here, or if types are just not a part of this library.

**Comment by mdmintz** [comment]
That depends on the [SeleniumBase Syntax Format](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md) used. With the `BaseCase` and `sb` `pytest` fixture formats, the IDE knows the base methods. In the case of the `SB()` and `Driver()` formats, the methods are determined at runtime, so if `uc=True`, then the UC Mode methods appear. If activating CDP Mode, then more methods appear.

When using the Python debugger, you'll see all the methods. Set a `breakpoint()`, to get realtime auto-completion for the methods.

This assumes you're looking for the methods, rather than the types associated.
For the types, that can be found in the `SB()` and `Driver()` docstrings. Eg:

```python
>>> from seleniumbase import Driver
>>> help(Driver) 
```

**Reply by albertkim** [reply]
I'm using `21. SeleniumBase SB (Python context manager)` since my use case is more scraping and less tests.

**Reply by mdmintz** [reply]
From Debug Mode:

```python
(Pdb+) from seleniumbase import SB
(Pdb+) help(SB)
```

Then scroll down...
See all methods. See all types.
